redis/src/types.rs

```diff
@@ -1351,6 +1351,68 @@ pub trait RedisWrite {
     /// Appends an empty argument to the command, and returns a
     /// [`std::io::Write`] instance that can write to it.
     fn writer_for_next_arg(&mut self) -> impl std::io::Write + '_;
+
+    #[cfg(feature = "bytes")]
+    /// Appends an empty argument to the command, and returns a
+    /// [`bytes::BufMut`] instance that can write to it.
+    ///
+    /// `capacity` should be equal or greater to the amount of bytes
+    /// expected, as some implementations might not be able to resize
+    /// the returned buffer.
+    fn bufmut_for_next_arg(&mut self, capacity: usize) -> impl bytes::BufMut + '_ {
```

Comment:
maybe this should return a concrete type / an extending trait to BufMut, that allows writing multiple args? That way the user doesn't have to pay the cost of reallocation for each arg.

Reply:
Maybe something like:
```rust
fn as_bufmut_writer(&mut self) -> impl RedisBufMutWriter + '_;

pub trait RedisBufMutWriter {
    // Reserve at least `additional` capacity.
    fn reserve(&mut self, additional: usize);
    // Get a `bytes::BufMut` to write an arg to.
    //
    // `size` is the expected size of the arg, if `size` is less than
    // the amount of bytes written the implementation is allowed to panic.
    fn write_arg(&mut self, size: usize) -> impl bytes::BufMut + '_;
}
```
Or we can add the `reserve` function to the RedisWriter implementation (with a default implementation that does nothing).

Reply:
How `RedisWriter::reserve` be implemented for `Vec<Vec<u8>>`?

Reply:
It can't and will be a no-op in that case, but that's fine as it's an optional optimisation. For `Vec<Vec<u8>>` you would need a `reserve_args(n: usize)` function or something similar.

Reply:
I would document `reserve` like this:
```rust
/// Tell the writer to expect at least `additional` more bytes
///
/// These additional bytes can be part of one or more arguments. The implementer is allowed to
/// ignore this information if it's internal command buffer is not continuous, like [`Vec<Vec<u8>>`]
fn reserve(&mut self, additional: usize);
```
And maybe rename `reserve` to something more fitting.

Reply:
```rust
    /// Reserve space for `additional` arguments in the command
    ///
    /// `additional` is a list of the byte sizes of the arguments.
    ///
    /// It's strongly recommended that implementors implement this function,
    /// as the default implementation is a no-op.
    fn reserve_space_for_args(&mut self, additional: &[usize])
```
I came up with this function for the RedisWriter trait. This allows both `Cmd` and `Vec<Vec<u8>>` to utilize the information. I've again provided a default implementation (a no-op), so that this isn't a breaking change for anyone.